### DISCUSSION TITLE
Testing app with a 'src' directory and getting all the paths to work

### DISCUSSION THREAD

**Discussion by boanergies** [fastapi/fastapi#8748] [Questions] [answered]
### First check

* [X] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

My project is organized with a 'src' directory to help with deploying, meaning it has the directory structure shown below:

```
project/src/
project/src/static
preojct/tests/
```

If I have a main.py in 'src' and a test_main.py in 'test' I have to run pytest with the following command from the root directory project to make sure all the paths work out for importing

`python -m pytest`

Main app file
```Python
from fastapi import FastAPI

app = FastAPI()
app.mount("/static", StaticFiles(directory="static"), name="static")

@app.get("/")
def read_root():
    return {"Hello": "World"}
```

Test File
```Python

from fastapi.testclient import TestClient

from src.main import app

client = TestClient(app)


def test_get():
    response = client.get("/")
    assert response.status_code == 200
```

### Description

* run `python -m pytest`
* Exception thrown during the import of app due to not being able to find the 'static' directory that is being mounted

So the real question is, how can you set the mounting correctly for testing since when I run pytest it's looking for the 'static' directory in project rather than src.

### Environment

* OS: Linux or Windows:
* FastAPI Version 0.66.0:
* Pytest Version 6.2.4:
* Python version: 3.9.5

**Comment by boanergies** [comment]
I tried adding a __init__.py to the project directory, and then changed 'static' to 'project/static' which helped with the testing but didn't help with the deployment of my project, specifically I'm using [deta](https://www.deta.sh/) for my project. Instead I got errors in deployment that it couldn't find 'project/static' essentially.

**Comment by boanergies** [comment]
Playing with this more I think the particular problem is focused more on how deta deploys code to the cloud. I've messaged them on slack and will how they think code should be organized

**Comment by ticdenis** [comment]
Just compute the realpath like this: `app.mount('/static', StaticFiles(directory=realpath(f'{realpath(__file__)}/../static')), name='static')`

**Comment by boanergies** [comment]
I'm not a fan of having created paths like that to vary between testing and deployment. Also my deployment would break due to how deta deploys code in production.

**Comment by jaxor24** [comment, marked as answer]
I was able to fix this by adding a utils module at the same level as the app main with a method to return the project root. This can then be used to locate static files, templates etc.

In the main.py:

```python
app = FastAPI()
app.mount("/static", StaticFiles(directory=utils.get_project_root() / "static"), name="static")
```

In the utils.py at the same level:

```python
def get_project_root():
    return Path(__file__).parent
```

**Comment by dmarcoux** [comment]
I have a similar solution to @jaxor24, but instead of `utils.py`, I rely on Starlette's `Config` with a `config.py` file at the same level of `main.py`.

It looks like this:

<details open>
<summary>config.py</summary>

```python
from pathlib import Path

from starlette.config import Config

# Config will read first from environment variables, then from the `.env` file at the git repository's root
config = Config(".env")

# Static files
DEFAULT_STATIC_DIR = Path(__file__).parent / "static"
STATIC_DIR = config("STATIC_DIR", default=DEFAULT_STATIC_DIR)
```

</details>

<details open>
<summary>main.py</summary>

```python
from YOUR_PYTHON_PACKAGE_NAME.config import STATIC_DIR

app = FastAPI()

app.mount("/static", StaticFiles(directory=STATIC_DIR), name="static")

(...)
```

</details>
